### PR TITLE
Add a ContentSecurityPolicy datastructure

### PR DESCRIPTION
**Pull request by pgjones**
This should help make CSP headers easier to construct and read, by
adding structure for the directives. It is based on today's version of
https://www.w3.org/TR/CSP3/ .

### CODE REVIEW

**Comment by davidism** [comment]
Nice! Is this something [Flask-Talisman](https://github.com/GoogleCloudPlatform/flask-talisman) can take advantage of too? Is there anything in their implementation that would be useful here?

File: src/werkzeug/datastructures.py [hunk 1]
```diff
@@ -2012,6 +2012,96 @@ class ResponseCacheControl(_CacheControl):
...
+        self.provided = values is not None
+
+    def _get_value(self, key):
+        """Used internally by the accessor properties."""
+        return self.get(key)
+
+    def _set_value(self, key, value):
+        """Used internally by the accessor properties."""
+        if value is None:
+            self.pop(key)
```
Comment (ThiefMaster) [inline src/werkzeug/datastructures.py]: I think setting a value to `None` twice will fail here. You probably want `self.pop(key, None)`

Comment (pgjones) [reply src/werkzeug/datastructures.py]: I followed the convention set [here](https://github.com/pallets/werkzeug/blob/master/src/werkzeug/datastructures.py#L1943) want me to change both?

Comment (ThiefMaster) [reply src/werkzeug/datastructures.py]: Yes, that code is broken, too:

```
In [17]: x.no_cache = True

In [18]: x.no_cache = False

In [19]: x.no_cache = False

In [20]: x.no_cache = None

In [21]: x.no_cache = None
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
```

Comment (pgjones) [reply src/werkzeug/datastructures.py]: ok, see #1618

**Comment by pgjones** [comment]
I think it could be used by Flask-Talisman. I quite like the ability (in Flask-Talisman) to do something like `csp.default_src = ["'self'", "something"]` i.e. set a list. I thought about doing this in Werkzeug, but the complexity starts to grow very quickly - not every directive-value is a list, then 'unsafe-inline' is only valid in some directives etc... So I think this is a good place to start.

**Comment by pgjones** [comment]
@davidism are you happy with the API? (I plan to add the same API to Quart and hope to get Werkzeug sign-off).

**Comment by davidism** [comment]
Overall I'm fine with the design. Flask-Talisman has a few CSP features that I'd at least like to think about now even if they can be implemented later.

* The `Content-Security-Policy-Report-Only` header.
* Adding a per-response nonce to CSP sections.

**Comment by davidism** [comment]
@theacodes We're implementing a CSP header in Werkzeug, would love to hear any feedback you have.

**Comment by theacodes** [comment]
The API here looks fine for having a structured way of setting the CSP header.

Talisman can certainly be updated to use this, but I'd also be happy to see the need for Talisman either diminish or evaporate completely.

**Comment by pgjones** [comment]
@davidism I've added the Report-Only variant. This already allows a nonce to be set (although it requires some knowledge about how to format it), I think anything beyond this is best tied in with Flask (so as to inject the nonce into the HTML).

**Comment by pgjones** [comment]
🎉 thanks.

**Comment by jezdez** [comment]
Woah, I had missed this, super exciting @pgjones @theacodes! Thank you ❤️

**Pull request merged**